### PR TITLE
Shrs Insulter

### PR DESCRIPTION
**Pull request by nithinmuthukumar**
# Content
New plugin which is a port of bash insulter

# Considerations
- Passing optional arguments to a state is awkward since it needs to pass through plugin, parameters are duplicated twice or state needs to be a field.
- Could style the text later?

### CODE REVIEW

**Comment by nithinmuthukumar** [comment]
Doesn't work for invalid commands only works for commands that fail currently

example: ls awrgb

File: plugins/shrs_insulter/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            insults: insults_c.iter().map(|i| i.to_string()).collect(),
+            freq,
+            rand: thread_rng(),
+        }
+    }
+
+    fn should_insult(&mut self) -> bool {
+        self.rand.gen::<f32>() <= self.freq
+    }
+    fn rand_insult(&mut self) -> &str {
```
Comment (MrPicklePinosaur) [inline plugins/shrs_insulter/src/lib.rs:95]: Semantically, rand_insult should not take a mutable reference to self (you don't get the image of mutating state when fetching something for a list). Optional, but you can wrap rand in a [RefCell](https://doc.rust-lang.org/std/cell/index.html) to introduce interior mutability.

**Review by MrPicklePinosaur — CHANGES_REQUESTED** [review]

**Comment by MrPicklePinosaur** [comment]
> Doesn't work for invalid commands only works for commands that fail currently
> 
> example: ls awrgb

hopefully new command lang refactor should expose specific types of errors (ie command not found) to make error types more distinct

File: plugins/shrs_insulter/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    }
+}
+pub struct InsulterPlugin {
+    insults: Vec<String>,
+    freq: f32,
+    include_default: bool,
+}
+
+impl InsulterPlugin {
+    pub fn new(insults: Vec<String>, freq: f32, include_default: bool) -> Self {
```
[the remark this answers is not in the document]
Comment (nithinmuthukumar) [reply plugins/shrs_insulter/src/lib.rs:113]: InsulterState doesn't need to implement default.

**Comment by nithinmuthukumar** [comment]
Also just learned that
 [rand::thread_rng() 9](https://docs.rs/rand/latest/rand/fn.thread_rng.html) already caches the ThreadRng object after the first call and retrieves it on later calls, so storing it yourself makes no difference.

**Pull request merged**